### PR TITLE
Fixes for "Perl in Space"

### PR DESCRIPTION
**Pull request by mohawk2**

### CODE REVIEW

File: t/alias.t [hunk 1]
```diff
@@ -25,7 +25,7 @@ do {
   close $fh;
 };
 
-eval { $config->set_alias("myecho1", "$^X $script_name f00f") };
+eval { $config->set_alias("myecho1", [ $^X, $script_name, "f00f" ]) };
```
Comment (plicease) [inline t/alias.t]: should test both string and array ref modes.

File: lib/Shell/Config/Generate.pm [hunk 2]
```diff
@@ -366,7 +366,7 @@ sub _value_escape_win32
 {
   my $value = shift() . '';
   $value =~ s/%/%%/g;
-  $value =~ s/([&^|<>()])/^$1/g;
+  $value =~ s/([&^|<>])/^$1/g;
```
Comment (plicease) [inline lib/Shell/Config/Generate.pm]: Can you elaborate on this?

Comment (mohawk2) [inline lib/Shell/Config/Generate.pm]: The directory with a space I have, also has parens in, similar to `Program Files (x86)`. If those get quoted, it seems the command/cmd interpreter don't like it.

File: t/alias.t (hunk 1 above)
Comment (mohawk2) [inline t/alias.t]: Adding this.

**Comment by plicease** [comment]
Fails with scalar usage on powershell:

https://gist.github.com/plicease/ac4e424c67076508d7374a1b6507bfe2

Spaces should be escaped on all shells in array mode, and should not be escaped in scalar/string mode.  There should be tests to make sure everything is escaped correctly (including special characters) in array mode.

The more I think about this feature I think I am a little wary of adding it.  I think it would be useful if implemented appropriately.  I think this PR should either:

 - implement the changes described above with tests
 - back out the array ref feature and skip the alias test when there is a space in $^X, add a caveat in the documentation

**Comment by mohawk2** [comment]
I went with option one, both modes now tested right and passing on my system. I respect and share your caution, but I think just skipping tests is a bad idea (at least without a better way of doing things) - it's not at all unusual on Windows to have executables with spaces in their paths!

File: lib/Shell/Config/Generate.pm [hunk 3]
```diff
@@ -631,7 +631,7 @@ sub generate
       elsif($shell->is_power)
       {
         # this leaves spaces between words, but words with spaces in get quoted
-        my $body = join ' ', map powershell_escape_path($_), @words;
+        my $body = (@words > 1) ? join(' ', map powershell_escape_path($_), @words) : $words[0];
```
Comment (plicease) [inline lib/Shell/Config/Generate.pm]: @words > 1 includes `['words']` and `'words'` should only incude `'words'`

**Comment by plicease** [comment]
That only handles powershell, not all the unix shells, cmd.exe etc.  There are no tests that make sure that special characters get escaped correctly.  Also this:

```perl
 $sch->alias( foo => 'single');
```

shouldn't be treated the same as

```perl
$sch->alias( foo => ['single'] );
```

I can see your point that programs can be installed in paths with spaces in them, but the usual use case for an alias is aliasing `cp=cp -i` or `ls=ls -CF` or something like that.  Given this is the most common use case, I think it is completely acceptable to say for this we don't handle escaping for aliases.  There are lots of skips in this distribution depending on the environment already, I think it is being tested in enough places that if the common use case is broken I will see it from cpantesters soon enough.

I would much prefer to mark this as a skip and open an issue for this to be implemented properly if/when someone actually needs it and if/when someone has the time to implement it properly.

**Comment by plicease** [comment]
I was able to use `Win32::GetShortPathName()` to solve the spaces in $^X problem for the alias test.  Which is good news we don't have to skip the alias test for perl in space.  I've opened  a new issue #17 which proposes the array reference mode.  I think it is not a bad idea, but implementation and testing for such a feature is more work right now than can be allocated to it.

**Pull request closed without merging**